shared-module/busdevice/I2CDevice.c

```diff
@@ -0,0 +1,88 @@
+/*
+ * This file is part of the MicroPython project, http://micropython.org/
+ *
+ * The MIT License (MIT)
+ *
+ * Copyright (c) 2020 Mark Komus
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+#include "shared-bindings/busdevice/I2CDevice.h"
+#include "shared-bindings/busio/I2C.h"
+#include "py/mperrno.h"
+#include "py/nlr.h"
+#include "py/runtime.h"
+
+void common_hal_busdevice_i2cdevice_construct(busdevice_i2cdevice_obj_t *self, busio_i2c_obj_t *i2c, uint8_t device_address, bool probe) {
+    self->i2c = i2c;
+    self->device_address = device_address;
+    self->probe = probe;
+}
+
+void common_hal_busdevice_i2cdevice_lock(busdevice_i2cdevice_obj_t *self) {
+    bool success = false;
+    while (!success) {
+        success = common_hal_busio_i2c_try_lock(self->i2c);
```

Comment:
The implementation in Python allows asynchronous exceptions like Ctrl-C to be processed here as well as allowing background tasks to run.  You may need to explicltly `RUN_BACKGROUND_TASKS; mp_handle_pending();` to allow both of these to take place as normal.

Reply:
@jepler I thought the equivalent of `mp_handle_pending()` was originally a part of running background tasks. If not, maybe we need to add it, since there are many places where these busy-wait loops should be interruptable.

Reply:
Maybe discuss this elsewhere

Reply:
I did try these two calls within the loop and it slowed the I2C calls down significantly. I'm not familiar enough with them to know why yet. Two thoughts are call them every X tries of the lock, or fail the lock after X tries with an exception.